controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,366 @@
+		if controllerutil.AddFinalizer(pool, infrav1.OpenStackFloatingIPPoolFinalizer) {
+			return ctrl.Result{Requeue: true}, r.Client.Update(context.Background(), pool)
+		}
+	} else {
+		// Handle deletion
+		return r.reconcileDelete(ctx, pool)
+	}
+
+	scope, err := r.ScopeFactory.NewClientScopeFromFloatingIPPool(ctx, r.Client, pool, r.CaCertificates, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.setIPStatuses(ctx, scope, pool); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	patchHelper, err := patch.NewHelper(pool, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	defer func() {
+		if err := patchHelper.Patch(ctx, pool); err != nil {
+			if reterr == nil {
+				reterr = fmt.Errorf("error patching OpenStackFloatingIPPool %s/%s: %w", pool.Namespace, pool.Name, err)
+			}
+		}
+	}()
+
+	claims := &ipamv1.IPAddressClaimList{}
+	if err := r.Client.List(context.Background(), claims, client.InNamespace(req.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, claim := range claims.Items {
+		claim := claim
+		log := log.WithValues("claim", claim.Name)
+
+		cluster, err := util.GetClusterFromMetadata(ctx, r.Client, claim.ObjectMeta)
+		if err != nil {
+			log.Info("Could not get cluster resource for IPAddressClaim")
+			return ctrl.Result{}, nil
+		}
+
+		infraCluster, err := r.getInfraCluster(ctx, cluster, &claim)
+		if err != nil {
+			return ctrl.Result{}, errors.New("error getting infra provider cluster")
+		} else if infraCluster == nil {
+			log.Info("infra cluster is not ready yet")
+			return ctrl.Result{}, nil
+		}
+
+		if claim.Status.AddressRef.Name == "" {
+			clusterName := fmt.Sprintf("%s-%s", claim.ObjectMeta.Labels[clusterv1.ClusterNameLabel], claim.Namespace)
+			ip, err := r.getIP(ctx, scope, pool, infraCluster, clusterName)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+
+			ipAddress := &ipamv1.IPAddress{
+				ObjectMeta: ctrl.ObjectMeta{
+					Name:      claim.Name,
+					Namespace: claim.Namespace,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: claim.APIVersion,
+							Kind:       claim.Kind,
+							Name:       claim.Name,
+							UID:        claim.UID,
+						},
+					},
+				},
+				Spec: ipamv1.IPAddressSpec{
+					ClaimRef: corev1.LocalObjectReference{
+						Name: claim.Name,
+					},
+					PoolRef: corev1.TypedLocalObjectReference{
+						APIGroup: pointer.String(infrav1.GroupVersion.Group),
+						Kind:     pool.Kind,
+						Name:     pool.Name,
+					},
+					Address: ip,
+				},
+			}
+
+			if !contains(pool.Spec.PreAllocatedFloatingIPs, ip) && pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete {
+				controllerutil.AddFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+			}
+
+			if err = r.Client.Create(ctx, ipAddress); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			claim.Status.AddressRef.Name = ipAddress.Name
```

Comment:
This could do with some robustification: we would never recover if we wrote the ipaddress but failed to update the claim. I think the simplest thing to do would be to attempt to fetch an ip address with the expected name first, and only call getIP and do the create if it doesn't exist already.

Reply:
I agree this is not robust enough, trying to fetching the ipamv1.IPAddress is probably a good start but that would not protect against orphan floating ips in openstack in case getIP allocates a new ip and the creation of the ipamv1.IPAddress fails.

I was thinking about throwing in a defer that would delete the IP in that case, but that felt a bit too hacky.. But atleast an orphaned fip would not break the next reconciliation

Reply:
We should jump here if IPAddress already existed above.